src/apipkg/__init__.py

```diff
@@ -48,33 +61,85 @@ def distribution_version(name):
+    """
     d = {}
-    f = getattr(oldmod, "__file__", None)
+    f = getattr(mod, "__file__", None)
     if f:
         f = _py_abspath(f)
     d["__file__"] = f
-    if hasattr(oldmod, "__version__"):
-        d["__version__"] = oldmod.__version__
-    if hasattr(oldmod, "__loader__"):
-        d["__loader__"] = oldmod.__loader__
-    if hasattr(oldmod, "__path__"):
-        d["__path__"] = [_py_abspath(p) for p in oldmod.__path__]
-    if hasattr(oldmod, "__package__"):
-        d["__package__"] = oldmod.__package__
-    if "__doc__" not in exportdefs and getattr(oldmod, "__doc__", None):
-        d["__doc__"] = oldmod.__doc__
-    d["__spec__"] = getattr(oldmod, "__spec__", None)
+    if hasattr(mod, "__version__"):
+        d["__version__"] = mod.__version__
+    if hasattr(mod, "__loader__"):
+        d["__loader__"] = mod.__loader__
+    if hasattr(mod, "__path__"):
+        d["__path__"] = [_py_abspath(p) for p in mod.__path__]
+    if hasattr(mod, "__package__"):
+        d["__package__"] = mod.__package__
+    if "__doc__" not in exportdefs and getattr(mod, "__doc__", None):
+        d["__doc__"] = mod.__doc__
+    d["__spec__"] = getattr(mod, "__spec__", None)
     d.update(attr)
-    if hasattr(oldmod, "__dict__"):
-        oldmod.__dict__.update(d)
+    if hasattr(mod, "__dict__"):
+        mod.__dict__.update(d)
     mod = ApiModule(pkgname, exportdefs, implprefix=pkgname, attr=d)
     sys.modules[pkgname] = mod
-    # eagerload in bypthon to avoid their monkeypatching breaking packages
-    if "bpython" in sys.modules or eager:
-        for module in list(sys.modules.values()):
-            if isinstance(module, ApiModule):
-                module.__dict__
+    return mod
+
+
+def _initpkg_py3(mod, pkgname, exportdefs, attr=None):
+    """Python 3 helper for initpkg.
+
+    Python 3.3+ uses finer grained locking for imports, and checks sys.modules before
+    acquiring the lock to avoid the overhead of the fine-grained locking. This
+    introduces a race condition when a module is imported by multiple threads
+    concurrently - some threads will see the initial module and some the replacement
+    ApiModule. We avoid this by updating the existing module in-place.
+
+    """
+    if mod is None:
+        d = {"__file__": None, "__spec__": None}
+        d.update(attr)
+        mod = ApiModule(pkgname, exportdefs, implprefix=pkgname, attr=d)
+        sys.modules[pkgname] = mod
+    else:
+        f = getattr(mod, "__file__", None)
+        if f:
+            f = _py_abspath(f)
+        mod.__file__ = f
+        if hasattr(mod, "__path__"):
+            mod.__path__ = [_py_abspath(p) for p in mod.__path__]
+        if "__doc__" in exportdefs and hasattr(mod, "__doc__"):
+            del mod.__doc__
+        for name in dir(mod):
+            if name not in _PRESERVED_MODULE_ATTRS:
+                delattr(mod, name)
+
+        # Updating class of existing module as per importlib.util.LazyLoader
+        mod.__class__ = ApiModule
```

Comment:
This breaks the compatibility with Python 3.4: https://github.com/pytest-dev/apipkg/issues/29.

Reply:
This is unfortunate, I'm not sure there's a way to fix the race condition if we can't update the module in place.

Options:
* Figure out a way to update the module in place in Python <= 3.4, maybe by copying from ApiModule into module instance?
* Accept the race condition in Python <= 3.4 and don't try to update in place
* Drop support for Python 3.4 completely, and update python_requires appropriately

I'm not sure what @RonnyPfannschmidt's intention was for 3.4 support, I just relied on the automated tests.

Reply:
I wonder if `setattr()` would somehow work.

Reply:
@GlenWalker dropping 3.4 from the matrix was accidental: https://github.com/pytest-dev/apipkg/issues/31#issue-1063182763

Reply:
> I wonder if setattr() would somehow work.

Setting attributes directly on the original module instance would probably work for `eager=True`, but not for `eager=False`. I had a quick look, but so far the only options for lazy loading require support from the module class (i.e. Python calls `__getattr__` from the module class, and not the module instance). We wouldn't want to update `__getattr__` on `types.ModuleType` globally, so if we can't update `__class__` on the module instance then we are stuck.

I suspect that for Python 3.4 we'll just have to accept the race condition, and replace the module, rather than updating it in place. Unless anyway has any good ideas?

(The race condition has existed for a long time without anyone complaining, so this might be acceptable)

Reply:
Sounds reasonable.